sodetlib/operations/bias_dets.py

```diff
@@ -201,3 +201,347 @@ def bias_to_rfrac(S, cfg, rfrac=0.5, bias_groups=None, iva=None,
+        drop_from_normal = False
+        for bl in bg_detectors_normal:
+            mask_bg = np.where(bsa_0.bgmap==bl)[0]
+            per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+            ## mask for normal
+            mask_normal = np.where((per_bl_percentage_rn_0 > 0.9) | (per_bl_percentage_rn_0 < -0.9))
+            ## if more than half of the detectors are normal
+            if len(mask_normal[0]) > 0.5*len(mask_bg):
+                drop_from_normal = True
+            else:
+                bg_detectors_normal.remove(bl)
+        for bl in bg_detectors_normal:
+            if previous_dc_biases_dfn[bl] < 0.5:
+                bg_detectors_normal.remove(bl)
+
+
+                
+
+    ##preparing for the 2nd step
+    initial_dc_biases = S.get_tes_bias_bipolar_array()
+    new_bias_voltage = initial_dc_biases.copy()
+
+
+
+
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+
+        per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+            
+        med_per_bl_percentage_rn = np.nanmedian(per_bl_percentage_rn_0)
+        
+        ## use the latest IV to decide the step size, might be tricky to get a good step size in field
+        v_sc = iva.v_bias[iva.idxs[[mask_bg], 0]]
+        v_norm = iva.v_bias[iva.idxs[[mask_bg], 1]]
+        v_spread = np.nanmedian(v_norm) - np.nanmedian(v_sc)
+        ## take 15% of the spread as the step size, so the next biasstep will be taken from 0.25 ~ 0.75 percent Rn after considering the spread
+        delta_dc_bias = 0.15 * v_spread
+        
+        
+        ## if the current bias point is below 50% Rn, increase V_biaa
+        if med_per_bl_percentage_rn < 0.5:
+            new_bias_voltage[bl] =  initial_dc_biases[bl] + delta_dc_bias
+        if med_per_bl_percentage_rn >= 0.5:
+            new_bias_voltage[bl] =  initial_dc_biases[bl] - delta_dc_bias
+
+    S.log(f"applying new voltage for 2nd biasstep {new_bias_voltage}")
+    S.set_tes_bias_bipolar_array(new_bias_voltage)  
+
+    ## taking the second bias step
+    S.log("taking the 2nd biasstep")
+    bsa_1 = bias_steps.take_bias_steps(S, cfg)
+    if not math_only:
+        bias_steps.plot_Rfrac(bsa_1)
+
+    ## check for script failue
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_1.bgmap==bl)[0]
+        per_bl_R0 = bsa_1.R0[mask_bg]
+        if np.isnan(np.nanmedian(per_bl_R0)):
+            ## the yield is very low, could it be the biasstep analysis script failed? repeat again
+            repeat_biasstep = True
+            S.log("the biasstep script might have failed, going to repeat the measurement")
+
+    if repeat_biasstep:
+        bsa_1 = bias_steps.take_bias_steps(S, cfg)
+        repeat_biasstep = False
+
+    percentage_rn_1 = bsa_1.R0/bsa_1.R_n_IV
+
+    target_percentage_rn_array = target_percentage_rn * np.ones(len(percentage_rn_0))
+    ## getting result from the previous 2
+    v0 = np.zeros(bsa_0.bgmap.shape[0])
+    for bl in range(12):
```

Comment:
Here and in line 403, use `bias_groups` and not `range(12)`

Reply:
range(12) is used such that the bias voltages in range(12) but not in bias_groups can be kept even that bias_line was not chosen to be re-biased. The 12 is the upper bar of how many biaslines SO UFMs can have.